mycok/user-directory
# 170677730: user-profile

#### What does this PR do

- Adds implementation for adding a user profile when creating or updating a user

#### Tasks that support your implementation

Requirements:

- Add create user json schema
- Add profile json schema
- install ajv as our json schema validator library
- Add a validate function to perform all the necessary json schema validations
- Add a custom error message function that returns the appropriate validation error messages

Testing:

- Add an invalid profile scenario outline
- Add new cucumber steps to support the invalid user profile scenario outline
- Add valid profile scenario outline
- Add reset db function to the create test steps
- Add a unit testing environment by installing mocha
- Add unit tests for the custom error message function
- Add custom eslint config for cucumber tests
- Add a general test script for both unit & e2e tests
- Add a unit test script
- Add unit tests for the validation Error class
- Add unit tests for check-content-type validation function
- Add unit tests for check-empty-payload validation function
- Add unit tests for check-content-type-is-json validation function
- Add unit tests for createUser controller function
- Add unit tests for the create engine
- Add integration tests for create engine

#### Changes made to the existing code if any

- Remove indepedent validation functions for required payload fields and their types
- Add password regex validation to the create json schema
- Modify the response messages to match the new validation error messages
- Isolate custom error message function into its own directory along with its unit tests
- Isolate validation error class into its own directory along with its unit tests
- Delete the custom password validation function
- Add pattern property to the create.json schema to validate the password with a regex
- Uncomment the password validation tests in both feature and steps files
- Isolate request header validation functions into their own directories along with their unit tests
- Rewrite the createUser controller function using the dependency injection approach
- Rewrite the injectControllerDepencies function to use the dependency injection approach
- Rewrite the create engine function to use the dependency injection approach
- Modify the user router module to add dependency injection functions and imports
- Fix the internal server error string typo in the create engine function
- Change the resolved promise returned value to return an entire result object instead of an id
- change the createUser controller resolved promise returned value to return an id

#### Tests recorded for this feature / fix




#### How can this solution be manually tested

##### Development Environment setup

- clone the repository and run `cd user-directory` to change to the project directory
- run `yarn install` command to install all the necessary dependencies
- run `yarn run dev` command to start the development server

##### Test Environment setup

- install & start `elasticsearch` from your terminal
- run `yarn test:server` command to start the testing server
- run `yarn test:e2e` command to run e2e tests
- run `yarn test:unit` command to run unit tests
- run `yarn test:integration` command to run integration tests
- run `yarn test` command to run all tests

#### Tickets/Stories related to this PR
[170677730](https://www.pivotaltracker.com/story/show/170677730)
